sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/EncryptionCosmosAsyncContainer.java

```diff
@@ -0,0 +1,353 @@
+            + "EncryptionContainer.");
+
+
+        byte[] streamPayload = cosmosSerializerToStream(item);
+        CosmosItemRequestOptions finalRequestOptions = requestOptions;
+        return this.encryptionProcessor.encrypt(streamPayload).flatMap(encryptedPayload -> this.container.createItem(
+            encryptedPayload,
+            partitionKey,
+            finalRequestOptions)
+            .publishOn(encryptionScheduler).flatMap(cosmosItemResponse -> setByteArrayContent(cosmosItemResponse,
+                this.encryptionProcessor.decrypt(EncryptionModelBridgeInternal.getByteArrayContent(cosmosItemResponse)))
+                .map(bytes -> this.responseFactory.createItemResponse(cosmosItemResponse, (Class<T>) item.getClass()))));
+    }
+
+
+    /**
+     * Deletes the item.
+     * <p>
+     * After subscription the operation will be performed. The {@link Mono} upon successful completion will contain a
+     * single Cosmos item response with the deleted item.
+     *
+     * @param itemId       id of the item.
+     * @param partitionKey partitionKey of the item.
+     * @param options      the request options.
+     * @return an {@link Mono} containing the Cosmos item resource response.
+     */
+    public Mono<CosmosItemResponse<Object>> deleteItem(String itemId,
+                                                       PartitionKey partitionKey,
+                                                       CosmosItemRequestOptions options) {
+
+        return container.deleteItem(itemId, partitionKey, options);
+    }
+
+    /**
+     * upserts item and encrypts the requested fields
+     *
+     * @param item           the Cosmos item represented as a POJO or Cosmos item object.
+     * @param partitionKey   the partition key.
+     * @param requestOptions request option
+     * @param <T>            serialization class type
+     * @return result
+     */
+    public <T> Mono<CosmosItemResponse<T>> upsertItem(T item,
+                                                      PartitionKey partitionKey,
+                                                      CosmosItemRequestOptions requestOptions) {
+        Preconditions.checkNotNull(item, "item");
+        if (requestOptions == null) {
+            requestOptions = new CosmosItemRequestOptions();
+        }
+
+        Preconditions.checkArgument(partitionKey != null, "partitionKey cannot be null for operations using "
+            + "EncryptionContainer.");
+
+
+        byte[] streamPayload = cosmosSerializerToStream(item);
+        CosmosItemRequestOptions finalRequestOptions = requestOptions;
+        return this.encryptionProcessor.encrypt(streamPayload).flatMap(encryptedPayload -> this.container.upsertItem(
+            encryptedPayload,
+            partitionKey,
+            finalRequestOptions)
+            .publishOn(encryptionScheduler).flatMap(cosmosItemResponse -> setByteArrayContent(cosmosItemResponse,
```

Comment:
So encryption happen before code goes into cosmos sdk layer . So encryption will happen on user's thread instead of cosmos netty io. Please share your thoughts